File: spec/system/creates_a_post_spec.rb
require 'rails_helper'

feature 'creates a post', type: :system, js: true do
  let!(:category) { create(:category) }

  context 'the post is valid' do
    it 'creates a post' do
      visit new_post_url

      fill_in 'post_title', with: 'my great post'
      fill_in 'post_body', with: 'body'
      select category.name, from: 'post[category_id]'
      click_on 'Create Post'

      Post.all.reload

      expect(Post.count).to eq 1
      expect(Post.first.title).to eq 'my great post'
      expect(page).to have_content('my great post')
    end
  end

  context 'the post in not vald' do
    it 'displays a flash message' do
      visit new_post_url

      fill_in 'post_title', with: 'a'
      fill_in 'post_body', with: 'body'
      select category.name, from: 'post[category_id]'
      click_on 'Create Post'

      Post.all.reload

      expect(Post.count).to eq 0
      expect(page).to have_content('too short')
    end
  end
end

File: cypress/integration/creates_a_post_spec.js
const context = describe

describe('Creates a post', () => {
  beforeEach(() => {
    cy.cleanDatabase()
  })

  context('the post is valid', () => {
    it('redirects to the created post', () => {
      cy.visit('localhost:3000/posts/new')

      cy.get('#post_title').type('my post', {force: true})
      cy.get('#post_body').type('this is the post body', {force: true})
      cy.get('#post_category_id').select('ruby', {force: true})

      cy.get('input[type="submit"]').click()

      cy.get('.category').contains('Category: ruby')
    })
  })

  context('post title is not valid', () => {
    it('shows flash message with error', () => {
      cy.visit('localhost:3000/posts/new')

      cy.get('#post_title').type('aa', {force: true})
      cy.get('#post_body').type('this is the post body', {force: true})
      cy.get('#post_category_id').select('ruby', {force: true})

      cy.get('input[type="submit"]').click()

      cy.get('body').contains('Title is too short')
    })
  })
})

describe('views posts index', () => {
  before(() => {
    cy.cleanDatabase({ seed: false }).then(() => {
      cy.seedPosts(5)
    })
  })

  it('views index page with some posts', () => {
    cy.visit('localhost:3000/posts')

    cy.get('.posts')
    cy.get('.posts').find('.post').should('have.length', 5)
  })
})
